tests/scipy_optimize_test.py

```diff
@@ -0,0 +1,285 @@
+      self.assert_wolfe(s, phi, derphi, err_msg="%s %g" % (name, old_phi0))
+
+  # -- Generic line searches
+
+  def test_line_search_wolfe2(self):
+    c = 0
+    smax = 512
+    for name, f, fprime, x, p, old_f in self.line_iter():
+      f0 = f(x)
+      g0 = fprime(x)
+      self.fcount = 0
+      res = line_search(value_and_grad(f, fprime), x, p, f_0=f0, g_0=g0)
+      s = res.a_k
+      fc = res.nfev
+      gc = res.ngev
+      fv = res.f_k
+      gv = res.g_k
+      # s, fc, gc, fv, ofv, gv = ls.line_search_wolfe2(f, fprime, x, p,
+      #                                                g0, f0, old_f,
+      #                                                amax=smax)
+      # assert_equal(self.fcount, fc+gc)
+      self.assertAllClose(fv, f(x + s * p), check_dtypes=False, atol=1e-5)
+      if gv is not None:
+        self.assertAllClose(gv, fprime(x + s * p), check_dtypes=False, atol=1e-5)
+      if s < smax:
+        c += 1
+    self.assertTrue(c > 3)  # check that the iterator really works...
+
+  def test_line_search_wolfe2_bounds(self):
+    # See gh-7475
+
+    # For this f and p, starting at a point on axis 0, the strong Wolfe
+    # condition 2 is met if and only if the step length s satisfies
+    # |x + s| <= c2 * |x|
+    f = lambda x: jnp.dot(x, x)
+    fp = lambda x: 2 * x
+    p = jnp.array([1, 0])
+
+    # Smallest s satisfying strong Wolfe conditions for these arguments is 30
+    x = -60 * p
+    c2 = 0.5
+
+    res = line_search(value_and_grad(f, fp), x, p, c2=c2)
+    s = res.a_k
+    # s, _, _, _, _, _ = ls.line_search_wolfe2(f, fp, x, p, amax=30, c2=c2)
+    self.assert_line_wolfe(x, p, s, f, fp)
+    self.assertTrue(s >= 30.)
+
+    res = line_search(value_and_grad(f, fp), x, p, c2=c2, max_iterations=5)
+    self.assertTrue(res.failed)
+    # s=30 will only be tried on the 6th iteration, so this won't converge
+
+  def test_line_search(self):
+    import jax
+
+    import jax.numpy as np
+
+    def f(x):
+      return np.cos(np.sum(np.exp(-x)) ** 2)
+
+    # assert not line_search(jax.value_and_grad(f), np.ones(2), np.array([-0.5, -0.25])).failed
+    xk = np.ones(2)
+    pk = np.array([-0.5, -0.25])
+    res = line_search(jax.value_and_grad(f), xk, pk, max_iterations=100)
+
+    from scipy.optimize.linesearch import line_search_wolfe2
+
+    scipy_res = line_search_wolfe2(f, jax.grad(f), xk, pk)
+
+    # print(scipy_res[0], res.a_k)
+    # print(scipy_res[3], res.f_k)
+
+    self.assertAllClose(scipy_res[0], res.a_k, atol=1e-5, check_dtypes=False)
+    self.assertAllClose(scipy_res[3], res.f_k, atol=1e-5, check_dtypes=False)
+
+  # -- More specific tests
+
+
+class TestBFGS(jtu.JaxTestCase):
+  def test_minimize(self):
+    from scipy.optimize import minimize as smin
+    import numpy as onp
+
+    def compare(func, x0):
+      # @jax.jit
+      def min_op(x0):
+        result = bfgs_minimize(func(jnp), x0,
+                               options=dict(ls_maxiter=10, maxiter=10, analytic_initial_hessian=False,
```

Comment:
Is `maxiter=10` enough to get all the way to convergence? For these simple functions, it might be close, but a better unit test would also be to compare results after only a few iterations.

I would suggest including tests similar to what I did for `scipy.sparse.linalg.cg`:

- `maxiter=1`
- `maxiter=3`
- no limit on `maxiter` -- go all the way to convergence

Reply:
Done using parametrized.

Reply:
Actually, this BFGS deviates from scipy's for the first few iterations. It's likely due to the line search. I realise now that `fmin_bfgs` uses `scipy.optimize.linesearch.line_search_wolfe12` which says,

```
Same as line_search_wolfe1 (which is actually  DCSRCH from MINPACK), but fall back to line_search_wolfe2 (which is the algorithm from Wright and Nocedal, 'Numerical Optimization', 1999) if suitable step length is not found.
```

I have implemented `line_search_wolfe2`. I need to read up on DCSRCH to understand the exact difference. However, both satisfy the strong Wolfe conditions when they converge.

Reply:
Hybrid DCSRCH/line_search_wolfe2 (as in scipy) not needed, so line search will not exactly match scipy.